packages/curve-ui-kit/src/features/connect-wallet/lib/ConnectionContext.tsx

```diff
@@ -0,0 +1,179 @@
+import { ethers } from 'ethers'
+import { createContext, type ReactNode, useContext, useEffect, useRef, useState } from 'react'
+import { getWalletChainId, getWalletSignerAddress, useSetChain, useWallet } from '@ui-kit/features/connect-wallet'
+import { WalletNameStorageKey } from '@ui-kit/features/connect-wallet/lib/hooks'
+import { getFromLocalStorage, setLocalStorage } from '@ui-kit/hooks/useLocalStorage'
+import { t } from '@ui-kit/lib/i18n'
+import { REFRESH_INTERVAL } from '@ui-kit/lib/model'
+import type { WalletState as Wallet } from '@web3-onboard/core'
+
+export type ConnectState = {
+  status: 'loading' | 'success' | 'failure' | ''
+  stage: 'api' | 'connect-wallet' | 'switch-network' | 'disconnect-wallet' | 'hydrate' | ''
+}
+
+export const CONNECT_STAGE = {
+  CONNECT_API: 'api',
+  HYDRATE: 'hydrate',
+  CONNECT_WALLET: 'connect-wallet',
+  DISCONNECT_WALLET: 'disconnect-wallet',
+  SWITCH_NETWORK: 'switch-network',
+} as const
+
+export const isSuccess = (connectState: ConnectState) => connectState.status === 'success'
+
+export const isFailure = ({ stage: connectionStage, status }: ConnectState, stage?: string) =>
+  stage ? status === 'failure' && connectionStage.startsWith(stage) : status === 'failure'
+
+export const isLoading = (connectState: ConnectState, stage?: string | string[]) =>
+  connectState.status === 'loading' &&
+  (!stage ||
+    (Array.isArray(stage) ? stage.some((s) => connectState.stage.startsWith(s)) : connectState.stage.startsWith(stage)))
+
+type ConnectionContextValue<TLib> = {
+  connectState: ConnectState
+  // todo: support multiple libs so we can share them between different apps
+  lib: TLib | null
+}
+
+const ConnectionContext = createContext<ConnectionContextValue<unknown>>({
+  connectState: { status: '', stage: '' },
+  lib: null,
+})
+
+const timeout = (message: string = t`Timeout connecting wallet`) =>
+  new Promise<never>((_, reject) => setTimeout(() => reject(new Error(message)), REFRESH_INTERVAL['3s']))
+
+// todo: this should be merged with apiStore and/or useWallet
```

Comment:
You say merge, but does this completely replace the apiStore? If I understand it correctly `libRef` is basically the singleton instance of the API, kinda like apiStore.

Reply:
I wanted to get away without any singleton.
I can try to use a store, I didn't because the provider hook was super useful to pass the initialization callbacks